File: temp/accumulator/accumulator.c
#include "accumulator.h"

#include <stdio.h>
#include <stdlib.h>

// matmul_base(
//     C : f32[16, 16] @DRAM,
//     A : f32[16, 16] @DRAM,
//     B : f32[16, 16] @DRAM
// )
void matmul_base( void *ctxt, float* C, const float* A, const float* B ) {
float acc;
for (int_fast32_t i = 0; i < 16; i++) {
  for (int_fast32_t j = 0; j < 16; j++) {
    acc = 0.0f;
    for (int_fast32_t k = 0; k < 16; k++) {
      acc += A[i * 16 + k] * B[k * 16 + j];
    }
    C[i * 16 + j] += acc;
  }
}
}

File: benchmarks/matmul/accumulator.py
from __future__ import annotations
from exo import *


@proc
def matmul_base(
    C: f32[16, 16] @ DRAM,
    A: f32[16, 16] @ DRAM,
    B: f32[16, 16] @ DRAM,
):
    acc: f32[1] @ DRAM
    for i in seq(0, 16):
        for j in seq(0, 16):
            acc[0] = 0.0
            for k in seq(0, 16):
                acc[0] += A[i, k] * B[k, j]

            C[i, j] += acc[0]
